src/nfloat/nfixed.c

```diff
@@ -1408,3 +1456,172 @@ _nfixed_mat_mul(nn_ptr C, nn_srcptr A, nn_srcptr B, slong m, slong n, slong p, s
     else
         _nfixed_mat_mul_classical(C, A, B, m, n, p, nlimbs);
 }
+
+/*
+  Given an m x n x p matrix multiplication with inputs bounded
+  entrywise by A, B and nlimbs precision:
+
+  - Set *bound* to a bound for the entries in all intermediate
+    variables of the computation. This should be < 1 to
+    ensure correctness.
+  - Set *error* to a bound for the output error, measured in ulp.
+
+   The caller can assume that the bound is a nondecreasing function
+   of A and B.
+
+   IMPORTANT: when changing the algorithm in _nfixed_mat_mul, this
+   must be changed to correspond.
+*/
+
+void
+_nfixed_mat_mul_bound_classical(double * bound, double * error, slong m, slong n, slong p, double A, double B, slong nlimbs)
+{
+    double fixup = 1.0 + 1e-6;
+
+    /* Error bound (in ulp) for naive scalar multiplication, and for dot product */
+    double error_mul = (2 * nlimbs - 1);
```

Comment:
Is this for high multiplication between two equally sized scalars?

Reply:
Yes. Note that this is the worst case bound for the sloppiest version of mulhigh, as used in ``_nfixed_dot_2``, ``_nfixed_dot_3``, ``_nfixed_dot_4``; one could use a tighter bound that depends on the actual algorithm used at a given ``nlimbs``.